File: C++/LinkedList/LoopDetection.cpp
/*
	@author
   	Aakash Verma
	
 	Detecting a loop in a linked list.

 	Output:
 	true

*/


#include <iostream>
using namespace std;

/* creating a node structure in C/C++ */
struct Node {
   int data;
   struct Node *next;
};

/* defining head as null as there is no node in the list initially. */
struct Node* head = NULL;

void push(int data) {
	Node *temp;
	temp = head;

	/* creating a node and assigning data to it and make its next as null */
	struct Node* new_node = (struct Node*) malloc(sizeof(struct Node));
	new_node->data = data;
   	new_node->next = NULL;

   	/* check if it's empty make new_node as head */
	if(head == NULL) {
   		head = new_node;
	}
	/* otherwise move upto the last and then connect last node with the new_node */
	else { 
		while(temp->next != NULL) {
			temp = temp->next;
		}
   		temp->next = new_node;
	}
}

/* finding loop in a linked list. */
bool isLoopFound() {
	Node *slow, *fast;
	slow = fast = head;

	if(head == NULL) {
		cout << "The list doesn't exist." << endl;
		return false;
	}

	while(fast != NULL && fast->next != NULL) {
		slow = slow->next;
		fast = fast->next->next;
		if(slow == fast) {
			return true;
		}
	}

	return false;
}

int main() {

	push(2);
	push(3);
	push(4);
	push(5);
	push(6);
	head->next->next->next->next->next = head->next->next;
	bool loopFound = isLoopFound();
	cout << loopFound << endl;
	return 0;

}

File: Python/LinkedList/LoopDetection.py
#
#
#
#   @author
#   Aakash Verma
#	
# 	Middle Of a Linked List
#
#	Output: 
#	true
#


# Below is the structute of a node which is used to create a new node every time.
class Node: 

	def __init__(self, data): 
		self.data = data
		self.next = None # None is nothing but null 

# Creating a class for implementing the code for Loop Detection in a LinkedList
class LinkedList: 
	
	# Whenever I'll create the object of a LinkedList class head will be pointing to null initially
	def __init__(self): 
		self.head = None


	# Inserting at the end of a Linked List (append function)
	def append(self, key): 

		h = self.head

		if h is None:
			new_node = Node(key)
			self.head = new_node
		else:
			while h.next != None:
				h = h.next
			new_node = Node(key)
			h.next = new_node

	# middle of a linked list
	def isLoopFound(self):
		fast = self.head
		slow = self.head

		if self.head is None:
			print("The list doesn't exist.")
			return False

		
		while fast is not None and fast.next is not None:
			slow = slow.next
			fast = fast.next.next
			if slow == fast:
				return True

		return False


# Code execution starts here 
if __name__=='__main__': 

	myList = LinkedList() 
	myList.append(3)
	myList.append(4)
	myList.append(5)
	myList.append(6)
	myList.append(7)
	myList.head.next.next.next.next.next = myList.head.next.next
	loopFound = myList.isLoopFound()
	print(loopFound)


